src/app/storage/FabricTableImpl.h

```diff
@@ -0,0 +1,140 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/support/TypeTraits.h>
+
+namespace chip {
+namespace app {
+namespace Storage {
+
+/**
+ * @brief Implementation of a storage in nonvolatile storage of a templatized table that stores by fabric id.
+ *
+ * FabricTableImpl is an implementation that allows to store arbitrary entities using PersistentStorageDelegate.
+ * It handles the storage of entities by their StorageId and EnpointID over multiple fabrics.
+ */
+template <class StorageId, class StorageData, size_t kIteratorsMax>
+class FabricTableImpl
+{
+    using TableEntry    = Data::TableEntry<StorageId, StorageData>;
+    using EntryIterator = CommonIterator<TableEntry>;
+    using EntryIndex    = Data::EntryIndex;
+
+public:
+    virtual ~FabricTableImpl() { Finish(); };
+
+    CHIP_ERROR Init(PersistentStorageDelegate * storage);
+    void Finish();
+
+    // Scene count
+    CHIP_ERROR GetEndpointEntryCount(uint8_t & entry_count);
+    CHIP_ERROR GetFabricEntryCount(FabricIndex fabric_index, uint8_t & entry_count);
+
+    // Data
+    CHIP_ERROR GetRemainingCapacity(FabricIndex fabric_index, uint8_t & capacity);
+    CHIP_ERROR SetTableEntry(FabricIndex fabric_index, const TableEntry & entry);
+    CHIP_ERROR GetTableEntry(FabricIndex fabric_index, StorageId entry_id, TableEntry & entry);
+    CHIP_ERROR RemoveTableEntry(FabricIndex fabric_index, StorageId entry_id);
+    CHIP_ERROR RemoveTableEntryAtPosition(EndpointId endpoint, FabricIndex fabric_index, EntryIndex entry_idx);
+
+    // Fabrics
+    CHIP_ERROR RemoveFabric(FabricIndex fabric_index);
+    CHIP_ERROR RemoveEndpoint();
+
+    void SetEndpoint(EndpointId endpoint);
+    void SetTableSize(uint16_t endpointEntryTableSize, uint16_t maxPerFabric);
+    bool IsInitialized() { return (mStorage != nullptr); }
+
+protected:
+    // gcc bug prevents us from using a static variable; see:
+    // https://stackoverflow.com/questions/50638053/constexpr-static-data-member-without-initializer
+    static constexpr EndpointId kEntryEndpointClusterID();
+
+    // This constructor is meant for test purposes, it allows to change the defined max for entries per fabric and global, which
+    // allows to simulate OTA where this value was changed
+    FabricTableImpl(uint16_t maxEntriesPerFabric, uint16_t maxEntriesPerEndpoint) :
+        mMaxPerFabric(maxEntriesPerFabric), mMaxPerEndpoint(maxEntriesPerEndpoint)
+    {}
+
+    // Global entry count
+    CHIP_ERROR SetEndpointEntryCount(const uint8_t & entry_count);
+
+    // wrapper function around emberAfGetClustersFromEndpoint to allow override when testing
+    virtual uint8_t GetClustersFromEndpoint(ClusterId * clusterList, uint8_t listLen);
+
+    // wrapper function around emberAfGetClusterCountForEndpoint to allow override when testing
+    virtual uint8_t GetClusterCountFromEndpoint();
+
+    class EntryIteratorImpl : public EntryIterator
```

Comment:
So to use one of these you just construct one wherever you want, right?  On the stack, for example?  MIght be worth documenting.